accessibility-insights-crawler is unavailable on npm

The accessibility-insights-scan package depends on accessibility-insights-crawler, but this package is not available in the npm registry. As a result, running ai-scan leads to a TypeError due to the missing ensureDirectoryFunc functionality in ai-scan-cli.js.

Steps to Reproduce:
Install accessibility-insights-scan:

Attempt to run a scan:
npx ai-scan --url https://www.wikipedia.org/ --output ./results

Observe the error:
TypeError: this.ensureDirectoryFunc is not a function

Check the npm registry for accessibility-insights-crawler:
npm install accessibility-insights-crawler

Result:
npm ERR! 404 Not Found - GET https://registry.npmjs.org/accessibility-insights-crawler


Expected Behavior:
The accessibility-insights-crawler package should be available in the npm registry, or the dependency on it should be removed or replaced with a built-in fallback in accessibility-insights-scan.

Observed Behavior:
The accessibility-insights-crawler package is missing from npm, causing runtime errors in accessibility-insights-scan.

Additional Context:
This issue blocks the ability to use accessibility-insights-scan effectively, as the dependency on a non-existent package leads to errors.

--- a/packages/cli/package.json
+++ b/packages/cli/package.json
@@ -1,6 +1,6 @@
 {
     "name": "accessibility-insights-scan",
-    "version": "3.0.1",
+    "version": "3.0.2",
     "description": "This project welcomes contributions and suggestions.  Most contributions require you to agree to a Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us the rights to use your contribution. For details, visit https://cla.microsoft.com.",
     "scripts": {
         "build": "webpack --config ./webpack.config.js \"$@\"",

--- a/packages/cli/src/axe/axe-info.ts
+++ b/packages/cli/src/axe/axe-info.ts
@@ -4,11 +4,11 @@
 import 'reflect-metadata';
 
 import * as Axe from 'axe-core';
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 
 @injectable()
 export class AxeInfo {
-    constructor(private readonly axe: typeof Axe = Axe) {}
+    constructor(@optional() @inject('Axe') private readonly axe: typeof Axe = Axe) {}
 
     public get version(): string {
         return this.axe.version;

--- a/packages/cli/src/baseline/baseline-file-updater.ts
+++ b/packages/cli/src/baseline/baseline-file-updater.ts
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { inject, injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { isEmpty } from 'lodash';
 import { OutputFileWriter } from '../files/output-file-writer';
 import { ScanArguments } from '../scan-arguments';
@@ -13,7 +13,7 @@ export class BaselineFileUpdater {
     constructor(
         @inject(BaselineFileFormatter) private readonly baselineFileFormatter: BaselineFileFormatter,
         @inject(OutputFileWriter) private readonly outputFileWriter: OutputFileWriter,
-        private readonly stdoutWriter: (output: string) => void = console.log,
+        @optional() @inject('stdoutWriter') private readonly stdoutWriter: (output: string) => void = console.log,
     ) {}
 
     public async updateBaseline(scanArguments: ScanArguments, baselineEvaluation: BaselineEvaluation): Promise<void> {

--- a/packages/cli/src/baseline/baseline-options-builder.ts
+++ b/packages/cli/src/baseline/baseline-options-builder.ts
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 import * as fs from 'fs';
-import { inject, injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { ScanArguments } from '../scan-arguments';
 import { BaselineFileFormatter } from './baseline-file-formatter';
 import { BaselineFileContent, BaselineOptions } from './baseline-types';
@@ -13,7 +13,7 @@ import { BaselineFileContent, BaselineOptions } from './baseline-types';
 export class BaselineOptionsBuilder {
     constructor(
         @inject(BaselineFileFormatter) private readonly baselineFileFormatter: BaselineFileFormatter,
-        private readonly fileSystem: typeof fs = fs,
+        @optional() @inject('fs') private readonly fileSystem: typeof fs = fs,
     ) {}
 
     public build(scanArguments: ScanArguments): BaselineOptions | null {

--- a/packages/cli/src/crawler/crawler-parameters-builder.ts
+++ b/packages/cli/src/crawler/crawler-parameters-builder.ts
@@ -2,13 +2,13 @@
 // Licensed under the MIT License.
 
 import * as fs from 'fs';
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { CrawlerRunOptions, Url } from 'accessibility-insights-crawler';
 import { ScanArguments } from '../scan-arguments';
 
 @injectable()
 export class CrawlerParametersBuilder {
-    constructor(private readonly urlObj: typeof Url = Url, private readonly fileSystem: typeof fs = fs) {}
+    constructor(@optional() @inject('fileSystem') private readonly urlObj: typeof Url = Url, private readonly fileSystem: typeof fs = fs) {}
 
     public build(scanArguments: ScanArguments): CrawlerRunOptions {
         if (scanArguments.crawl && scanArguments.url) {

--- a/packages/cli/src/files/output-file-writer.ts
+++ b/packages/cli/src/files/output-file-writer.ts
@@ -4,7 +4,7 @@
 import * as fs from 'fs';
 import path from 'path';
 import filenamifyCombined from 'filenamify';
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import normalizePath from 'normalize-path';
 import { ensureDirectory } from 'accessibility-insights-crawler';
 
@@ -16,9 +16,9 @@ const filenamifyUrl = require('filenamify-url');
 @injectable()
 export class OutputFileWriter {
     constructor(
-        private readonly fileSystem: typeof fs = fs,
-        private readonly fileSystemPath = path,
-        private readonly ensureDirectoryFunc: typeof ensureDirectory = ensureDirectory,
+        @optional() @inject('fs') private readonly fileSystem: typeof fs = fs,
+        @optional() @inject('fileSystemPath') private readonly fileSystemPath = path,
+        @optional() @inject('ensureDirectory') private readonly ensureDirectoryFunc: typeof ensureDirectory = ensureDirectory,
     ) {}
 
     public writeToDirectory(directory: string, fileBaseName: string, fileExtension: string, content: string): string {

--- a/packages/cli/src/runner/crawler-command-runner.ts
+++ b/packages/cli/src/runner/crawler-command-runner.ts
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 import * as fs from 'fs';
-import { inject, injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { isEmpty } from 'lodash';
 import { ScanArguments } from '../scan-arguments';
 import { ConsolidatedReportGenerator } from '../report/consolidated-report-generator';
@@ -24,8 +24,8 @@ export class CrawlerCommandRunner implements CommandRunner {
         @inject(BaselineOptionsBuilder) private readonly baselineOptionsBuilder: BaselineOptionsBuilder,
         @inject(BaselineFileUpdater) private readonly baselineFileUpdater: BaselineFileUpdater,
         @inject(ReportNameGenerator) private readonly reportNameGenerator: ReportNameGenerator,
-        private readonly fileSystem: typeof fs = fs,
-        private readonly stdoutWriter: (output: string) => void = console.log,
+        @optional() @inject('fs') private readonly fileSystem: typeof fs = fs,
+        @optional() @inject('stdoutWriter') private readonly stdoutWriter: (output: string) => void = console.log,
     ) {}
 
     public async runCommand(scanArguments: ScanArguments): Promise<void> {

--- a/packages/crawler/src/apify/apify-request-queue-creator.ts
+++ b/packages/crawler/src/apify/apify-request-queue-creator.ts
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 import * as fs from 'fs';
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import * as Crawlee from '@crawlee/puppeteer';
 import { ApifySettingsHandler, apifySettingsHandler } from './apify-settings';
 
@@ -25,8 +25,8 @@ export class ApifyRequestQueueCreator implements ResourceCreator {
     private readonly requestQueueName = 'scanRequests';
 
     public constructor(
-        private readonly settingsHandler: ApifySettingsHandler = apifySettingsHandler,
-        private readonly fileSystem: typeof fs = fs,
+        @optional() @inject('ApifySettingsHandler') private readonly settingsHandler: ApifySettingsHandler = apifySettingsHandler,
+        @optional() @inject('fs') private readonly fileSystem: typeof fs = fs,
     ) {}
 
     public async createRequestQueue(baseUrl: string, options?: RequestQueueOptions): Promise<Crawlee.RequestQueue> {

--- a/packages/crawler/src/common/hash-generator.ts
+++ b/packages/crawler/src/common/hash-generator.ts
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 import fnv1a from '@sindresorhus/fnv1a';
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { JumpConsistentHash } from './jump-consistent-hash';
 
 /* eslint-disable @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires */
@@ -10,7 +10,7 @@ const shaJS = require('sha.js');
 
 @injectable()
 export class HashGenerator {
-    public constructor(private readonly shaObj = shaJS) {}
+    public constructor(@optional() @inject('sha') private readonly shaObj = shaJS) {}
 
     public getWebsiteScanResultDocumentId(baseUrl: string, scanGroupId: string): string {
         // Preserve parameters order below for the hash generation compatibility

--- a/packages/crawler/src/common/promise-utils.ts
+++ b/packages/crawler/src/common/promise-utils.ts
@@ -1,11 +1,11 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 
 @injectable()
 export class PromiseUtils {
-    constructor(private readonly globalObj = global) {}
+    constructor(@optional() @inject('global') private readonly globalObj = global) {}
 
     public async waitFor<T, Y>(fn: Promise<T>, timeoutInMsec: number, onTimeoutCallback: () => Promise<Y>): Promise<T | Y> {
         let timeoutHandle: NodeJS.Timeout;

--- a/packages/crawler/src/crawler/crawler-configuration.ts
+++ b/packages/crawler/src/crawler/crawler-configuration.ts
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { isEmpty } from 'lodash';
 import { ApifySettings, ApifySettingsHandler, apifySettingsHandler } from '../apify/apify-settings';
 import { CrawlerRunOptions } from '../types/crawler-run-options';
@@ -13,7 +13,9 @@ export class CrawlerConfiguration {
     private _crawlerRunOptions: CrawlerRunOptions;
 
     public constructor(
-        private readonly settingsHandler: ApifySettingsHandler = apifySettingsHandler,
+        @optional() @inject('ApifySettingsHandler') private readonly settingsHandler: ApifySettingsHandler = apifySettingsHandler,
+        @optional()
+        @inject('DiscoveryPatternFactory')
         private readonly createDiscoveryPattern: DiscoveryPatternFactory = getDiscoveryPatternForUrl,
     ) {}
 

--- a/packages/crawler/src/crawler/site-crawler-engine.ts
+++ b/packages/crawler/src/crawler/site-crawler-engine.ts
@@ -3,7 +3,7 @@
 
 import * as fs from 'fs';
 import * as Crawlee from '@crawlee/puppeteer';
-import { inject, injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { isEmpty } from 'lodash';
 import * as Puppeteer from 'puppeteer';
 import { AuthenticatorFactory } from '../authenticator/authenticator-factory';
@@ -26,7 +26,7 @@ export class SiteCrawlerEngine implements CrawlerEngine {
         @inject(CrawlerFactory) private readonly crawlerFactory: CrawlerFactory,
         @inject(AuthenticatorFactory) private readonly authenticatorFactory: AuthenticatorFactory,
         @inject(CrawlerConfiguration) private readonly crawlerConfiguration: CrawlerConfiguration,
-        private readonly puppeteer: typeof Puppeteer = Puppeteer,
+        @optional() @inject('Puppeteer') private readonly puppeteer: typeof Puppeteer = Puppeteer,
     ) {}
 
     public async start(crawlerRunOptions: CrawlerRunOptions): Promise<void> {

--- a/packages/crawler/src/level-storage/data-base.ts
+++ b/packages/crawler/src/level-storage/data-base.ts
@@ -20,10 +20,10 @@ export class DataBase implements AsyncIterable<ScanResult> {
     private iterator: AsyncIterableIterator<string | Buffer>;
 
     constructor(
-        @inject(crawlerIocTypes.LevelUp) @optional() protected db?: LevelUp,
-        protected readonly levelupObj: typeof levelup = levelup,
-        protected readonly leveldownObj: typeof leveldown = leveldown,
-        protected readonly encodeObj: typeof encode = encode,
+        @optional() @inject(crawlerIocTypes.LevelUp) protected db?: LevelUp,
+        @optional() @inject('levelup') protected readonly levelupObj: typeof levelup = levelup,
+        @optional() @inject('leveldown') protected readonly leveldownObj: typeof leveldown = leveldown,
+        @optional() @inject('encode') protected readonly encodeObj: typeof encode = encode,
     ) {}
 
     public [Symbol.asyncIterator](): AsyncIterator<ScanResult> {

--- a/packages/crawler/src/logger/apify-console-logger-client.ts
+++ b/packages/crawler/src/logger/apify-console-logger-client.ts
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { Log } from '@apify/log';
 import moment from 'moment';
 import { LogLevel, LoggerClient, LoggerProperties } from './logger';
@@ -12,7 +12,7 @@ import { LogLevel, LoggerClient, LoggerProperties } from './logger';
 export class ApifyConsoleLoggerClient implements LoggerClient {
     private baseProperties: LoggerProperties;
 
-    constructor(private readonly logger: Log = new Log({ prefix: 'Scanner' })) {}
+    constructor(@optional() @inject('Log') private readonly logger: Log = new Log({ prefix: 'Scanner' })) {}
 
     public log(message: string, logLevel: LogLevel, properties?: LoggerProperties): void {
         const logMessage = `[${moment.utc().toISOString()}] ${message}`;

--- a/packages/crawler/src/logger/logger.ts
+++ b/packages/crawler/src/logger/logger.ts
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { ApifyConsoleLoggerClient } from './apify-console-logger-client';
 
 export enum LogLevel {
@@ -22,7 +22,7 @@ export interface LoggerClient {
 
 @injectable()
 export class Logger {
-    constructor(private readonly loggerClients: LoggerClient[] = [new ApifyConsoleLoggerClient()]) {}
+    constructor(@optional() @inject('LoggerClient') private readonly loggerClients: LoggerClient[] = [new ApifyConsoleLoggerClient()]) {}
 
     public setCommonProperties(properties: LoggerProperties): void {
         this.invokeLoggerClients((client) => client.setCommonProperties(properties));

--- a/packages/crawler/src/page-handler/page-handler.spec.ts
+++ b/packages/crawler/src/page-handler/page-handler.spec.ts
@@ -10,6 +10,8 @@ import { System } from '../common/system';
 import { PageHandler } from './page-handler';
 import { scrollToBottom } from './page-client-lib';
 
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
 type Writeable<T> = { -readonly [P in keyof T]: Writeable<T[P]> };
 
 describe(PageHandler, () => {
@@ -48,7 +50,9 @@ describe(PageHandler, () => {
         pageMock.setup((o) => o.isClosed()).returns(() => false);
         pageMock.setup((o) => o.url()).returns(() => 'url');
 
-        pageHandler = new PageHandler(loggerMock.object, checkIntervalMsecs, pageDomStableTimeMsec, scrollToBottomMock);
+        pageHandler = new PageHandler(loggerMock.object, scrollToBottomMock);
+        (pageHandler as any).pageDomStableTimeMsec = pageDomStableTimeMsec;
+        (pageHandler as any).checkIntervalMsecs = checkIntervalMsecs;
     });
 
     afterEach(() => {

--- a/packages/crawler/src/page-handler/page-handler.ts
+++ b/packages/crawler/src/page-handler/page-handler.ts
@@ -10,11 +10,13 @@ import { scrollToBottom } from './page-client-lib';
 
 @injectable()
 export class PageHandler {
+    private readonly checkIntervalMsecs: number = 200;
+
+    private readonly pageDomStableTimeMsec: number = puppeteerTimeoutConfig.pageDomStableTimeMsec;
+
     constructor(
-        @inject(Logger) @optional() private readonly logger: Logger,
-        private readonly checkIntervalMsecs: number = 200,
-        private readonly pageDomStableTimeMsec: number = puppeteerTimeoutConfig.pageDomStableTimeMsec,
-        private readonly scrollToPageBottom: typeof scrollToBottom = scrollToBottom,
+        @optional() @inject(Logger) private readonly logger: Logger,
+        @optional() @inject('scrollToBottom') private readonly scrollToPageBottom: typeof scrollToBottom = scrollToBottom,
     ) {}
 
     public async waitForPageToCompleteRendering(

--- a/packages/crawler/src/page-handler/page-navigation-hooks.ts
+++ b/packages/crawler/src/page-handler/page-navigation-hooks.ts
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { inject, injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { isNil } from 'lodash';
 import * as Puppeteer from 'puppeteer';
 import { BrowserError } from './browser-error';
@@ -16,7 +16,9 @@ export class PageNavigationHooks {
         @inject(PageConfigurator) public readonly pageConfigurator: PageConfigurator,
         @inject(PageResponseProcessor) protected readonly pageResponseProcessor: PageResponseProcessor,
         @inject(PageHandler) protected readonly pageRenderingHandler: PageHandler,
-        private readonly scrollTimeoutMsec = puppeteerTimeoutConfig.scrollTimeoutMsec,
+        @optional() @inject('scrollTimeoutMsec') private readonly scrollTimeoutMsec = puppeteerTimeoutConfig.scrollTimeoutMsec,
+        @optional()
+        @inject('pageRenderingTimeoutMsec')
         private readonly pageRenderingTimeoutMsec = puppeteerTimeoutConfig.pageRenderingTimeoutMsec,
     ) {}
 

--- a/packages/crawler/src/page-handler/page-response-processor.spec.ts
+++ b/packages/crawler/src/page-handler/page-response-processor.spec.ts
@@ -9,6 +9,7 @@ import { BrowserErrorTypes } from './browser-error';
 import { PageResponseProcessor } from './page-response-processor';
 
 /* eslint-disable @typescript-eslint/consistent-type-assertions */
+/* eslint-disable @typescript-eslint/no-explicit-any */
 
 describe(PageResponseProcessor, () => {
     let pageResponseProcessor: PageResponseProcessor;
@@ -138,7 +139,8 @@ describe('handles navigation errors', () => {
     ];
 
     test.each(testCaseMappings)('should parse navigation error: %o', async (testCase: NavigationErrorTestCase) => {
-        const pageResponseProcessor = new PageResponseProcessor(stubPatterns);
+        const pageResponseProcessor = new PageResponseProcessor();
+        (pageResponseProcessor as any).navigationErrorPatterns = stubPatterns;
 
         const actualError = pageResponseProcessor.getNavigationError({
             message: testCase.message,

--- a/packages/crawler/src/page-handler/page-response-processor.ts
+++ b/packages/crawler/src/page-handler/page-response-processor.ts
@@ -7,7 +7,7 @@ import { BrowserError, BrowserErrorTypes, pageNavigationErrorPatterns } from './
 
 @injectable()
 export class PageResponseProcessor {
-    constructor(private readonly navigationErrorPatterns: Partial<Record<BrowserErrorTypes, string[]>> = pageNavigationErrorPatterns) {}
+    private readonly navigationErrorPatterns: Partial<Record<BrowserErrorTypes, string[]>> = pageNavigationErrorPatterns;
 
     public getResponseError(response: Puppeteer.HTTPResponse, error: Error = new Error()): BrowserError {
         if (response.ok() !== true && response.status() !== 304) {

--- a/packages/crawler/src/page-processors/page-processor-base.ts
+++ b/packages/crawler/src/page-processors/page-processor-base.ts
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 import urlLib from 'url';
-import { inject, injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import * as Puppeteer from 'puppeteer';
 import { isArray } from 'lodash';
 import * as Crawlee from '@crawlee/puppeteer';
@@ -128,6 +128,8 @@ export abstract class PageProcessorBase implements PageProcessor {
         @inject(CrawlerConfiguration) protected readonly crawlerConfiguration: CrawlerConfiguration,
         @inject(PageNavigator) protected readonly pageNavigator: PageNavigator,
         @inject(Logger) protected readonly logger: Logger,
+        @optional()
+        @inject('saveSnapshot')
         protected readonly saveSnapshotExt: typeof Crawlee.puppeteerUtils.saveSnapshot = Crawlee.puppeteerUtils.saveSnapshot,
     ) {
         this.baseUrl = this.crawlerConfiguration.baseUrl();

--- a/packages/crawler/src/page-processors/simulator-page-processor.ts
+++ b/packages/crawler/src/page-processors/simulator-page-processor.ts
@@ -2,7 +2,7 @@
 // Licensed under the MIT License.
 
 import * as Crawlee from '@crawlee/puppeteer';
-import { inject, injectable } from 'inversify';
+import { inject, injectable, optional } from 'inversify';
 import { ActiveElement } from '../active-elements-finder';
 import { CrawlerConfiguration } from '../crawler/crawler-configuration';
 import { DataBase } from '../level-storage/data-base';
@@ -88,6 +88,8 @@ export class SimulatorPageProcessor extends PageProcessorBase {
         @inject(CrawlerConfiguration) protected readonly crawlerConfiguration: CrawlerConfiguration,
         @inject(PageNavigator) protected readonly pageNavigator: PageNavigator,
         @inject(Logger) protected readonly logger: Logger,
+        @optional()
+        @inject('saveSnapshot')
         protected readonly saveSnapshotExt: typeof Crawlee.puppeteerUtils.saveSnapshot = Crawlee.puppeteerUtils.saveSnapshot,
     ) {
         super(accessibilityScanOp, dataStore, blobStore, dataBase, crawlerConfiguration, pageNavigator, logger, saveSnapshotExt);

